lib/src/puppetdb-cli.cc

```diff
@@ -155,19 +159,45 @@ size_t write_body(char *ptr, size_t size, size_t nmemb, void *userdata){
     return size * nmemb;
 }
 
+string
+convert_query_to_post_data(const string& query_str) {
+    // If this is PQL then we need to wrap the query in double-quotes, otherwise
+    // the query is AST and we leave it alone
+    rapidjson::StringBuffer s;
+    rapidjson::Writer<rapidjson::StringBuffer> writer(s);
+    writer.StartObject();
+    writer.String("query");
+
+    string query_str_copy { query_str };
+    boost::trim(query_str_copy);
+    if (boost::starts_with(query_str_copy, "[")) {
+        rapidjson::StringStream is(query_str.c_str());
+        rapidjson::Reader reader;
+        if (!reader.Parse<rapidjson::kParseValidateEncodingFlag>(is, writer))
+            throw std::runtime_error { "Failed to parse query. Ensure your query is valid JSON." };
+    } else {
+        writer.String(query_str.c_str());
+    }
+
+    writer.EndObject();
+    return s.GetString();
+}
+
 void
 pdb_query(const PuppetDBConn& conn,
           const string& query_str) {
     auto curl = conn.getCurlHandle();
-    auto url_encoded_query = unique_ptr< char, function<void(char*)> >(
-        curl_easy_escape(curl.get(), query_str.c_str(), query_str.length()),
-        curl_free);
-
-    const auto server_url = conn.getServerUrl()
-            + "/pdb/query/v4?query="
-            + url_encoded_query.get();
+    const auto server_url = conn.getServerUrl() + "/pdb/query/v4";
 
     curl_easy_setopt(curl.get(), CURLOPT_URL, server_url.c_str());
+
+    const string post_data = convert_query_to_post_data(query_str);
```

Comment:
[1/4] Why memory is hard in C

Talked with @ajroetker over some of the specifics of this convert_to_post_data call. I think it's a bit of a thorny issue for non-C/C++ developers, writing up my understanding here in hopes of not stubbing our toe on it in the future.

The goal here is to take query_str parse/emit it with escaped values. The query_str variable is a const reference to string that is passed in (this becomes important later).